include/grub/slr_table.h

```diff
@@ -218,21 +214,50 @@ struct grub_slr_uefi_cfg_entry
   char evt_info[GRUB_TPM_EVENT_INFO_LENGTH];
 } GRUB_PACKED;
 
-static inline void *
+struct grub_slr_entry_uefi_config
+{
+  struct grub_slr_entry_hdr hdr;
+  grub_uint16_t revision;
+  grub_uint16_t nr_entries;
+  struct grub_slr_uefi_cfg_entry uefi_cfg_entries[];
+} GRUB_PACKED;
+
+static inline grub_addr_t
 grub_slr_end_of_entries (struct grub_slr_table *table)
 {
-  return (grub_uint8_t *) table + table->size;
+  return (grub_addr_t) table + table->size;
 }
 
 static inline struct grub_slr_entry_hdr *
 grub_slr_next_entry (struct grub_slr_table *table,
                      struct grub_slr_entry_hdr *curr)
 {
-  struct grub_slr_entry_hdr *next = (struct grub_slr_entry_hdr *)
-                                    ((grub_uint8_t *) curr + curr->size);
+  grub_addr_t addr;
+  struct grub_slr_entry_hdr *next;
+
+  /* Can read the size field of current entry? */
+  if ( grub_add ((grub_addr_t) curr, sizeof(*curr), &addr) )
+    return NULL;
```

Comment:
I suggest having a `grub_slr_first_entry()` function that validates that the first entry is valid.  Then `grub_slr_next_entry()` can either assume or assert that `curr` is valid.  By mathematical induction, this is enough to validate all entries in the table.

Here, this check is redundant, but you need to check that first `table->size` and then `curr->size` are at least `sizeof(*cur)`.

```suggestion
  /* Can read the size field of current entry? */
  if ( table->size < sizeof(*curr) )
    return NULL;

 /* Is the current entry large enough? */
 if ( curr->size < sizeof(*curr) )
   return NULL;
```

Reply:
But validity of the first entry is ensured by `grub_slr_init_table()` and code in `grub_slr_add_entry()`, what would `grub_slr_first_entry()` give on top of that?

Reply:
Nothing, consider this resolved.